# CMU-MOSI Multimodal SDK

# Download and Process mosi

### word level align

In [1]:
import mmsdk
from mmsdk import mmdatasdk
import numpy

In [2]:
#A simple averaging technique. More advanced methods can be built based on intervals.
def myavg(intervals,features):
        return numpy.average(features,axis=0)

In [3]:
#Downloading the dataset
# cmumosi_highlevel=mmdatasdk.mmdataset(mmdatasdk.cmu_mosi.highlevel,'cmumosi/')
cmumosi_highlevel=mmdatasdk.mmdataset('cmumosi/')

  0%|          | 0/93 [00:00<?, ? Computational Sequence Entries/s]

[2021-10-14 09:09:55.292] | Success | Computational sequence read from file cmumosi/CMU_MOSI_Visual_OpenFace_1.csd ...
[2021-10-14 09:09:55.316] | Status  | Checking the integrity of the <OpenFace_1> computational sequence ...
[2021-10-14 09:09:55.316] | Status  | Checking the format of the data in <OpenFace_1> computational sequence ...
[2021-10-14 09:09:55.376] | Success | <OpenFace_1> computational sequence data in correct format.
[2021-10-14 09:09:55.376] | Status  | Checking the format of the metadata in <OpenFace_1> computational sequence ...
[2021-10-14 09:09:55.376] | Warning | <OpenFace_1> computational sequence does not have all the required metadata ... continuing 
[2021-10-14 09:09:55.377] | Success | Computational sequence read from file cmumosi/CMU_MOSI_Visual_Facet_42.csd ...
[2021-10-14 09:09:55.398] | Status  | Checking the integrity of the <FACET_4.2> computational sequence ...
[2021-10-14 09:09:55.398] | Status  | Checking the format of the data in <FACET_4.2> comput

  0%|          | 0/93 [00:00<?, ? Computational Sequence Entries/s]

[2021-10-14 09:09:55.564] | Success | <OpenSmile_emobase2010> computational sequence data in correct format.
[2021-10-14 09:09:55.564] | Status  | Checking the format of the metadata in <OpenSmile_emobase2010> computational sequence ...
[2021-10-14 09:09:55.564] | Warning | <OpenSmile_emobase2010> computational sequence does not have all the required metadata ... continuing 
[2021-10-14 09:09:55.566] | Success | Computational sequence read from file cmumosi/CMU_MOSI_TimestampedWordVectors.csd ...
[2021-10-14 09:09:55.582] | Status  | Checking the integrity of the <glove_vectors> computational sequence ...
[2021-10-14 09:09:55.582] | Status  | Checking the format of the data in <glove_vectors> computational sequence ...
[2021-10-14 09:09:55.640] | Success | <glove_vectors> computational sequence data in correct format.
[2021-10-14 09:09:55.641] | Status  | Checking the format of the metadata in <glove_vectors> computational sequence ...
[2021-10-14 09:09:55.641] | Warning | <glove_vecto

  0%|          | 0/93 [00:00<?, ? Computational Sequence Entries/s]

[2021-10-14 09:09:55.803] | Success | <COVAREP> computational sequence data in correct format.
[2021-10-14 09:09:55.803] | Status  | Checking the format of the metadata in <COVAREP> computational sequence ...
[2021-10-14 09:09:55.803] | Warning | <COVAREP> computational sequence does not have all the required metadata ... continuing 
[2021-10-14 09:09:55.806] | Success | Computational sequence read from file cmumosi/CMU_MOSI_openSMILE_IS09.csd ...
[2021-10-14 09:09:55.826] | Status  | Checking the integrity of the <b'OpenSMILE'> computational sequence ...
[2021-10-14 09:09:55.826] | Status  | Checking the format of the data in <b'OpenSMILE'> computational sequence ...
[2021-10-14 09:09:55.875] | Success | <b'OpenSMILE'> computational sequence data in correct format.
[2021-10-14 09:09:55.875] | Status  | Checking the format of the metadata in <b'OpenSMILE'> computational sequence ...
[2021-10-14 09:09:55.875] | Warning | <b'OpenSMILE'> computational sequence does not have all the requir

[2021-10-14 09:09:56.015] | Success | <Opinion Segment Labels> computational sequence data in correct format.
[2021-10-14 09:09:56.015] | Status  | Checking the format of the metadata in <Opinion Segment Labels> computational sequence ...
[2021-10-14 09:09:56.015] | Warning | <Opinion Segment Labels> computational sequence does not have all the required metadata ... continuing 
[2021-10-14 09:09:56.015] | Success | Dataset initialized successfully ... 


In [9]:
#some random video from cmumosi_highlevel
some_video=list(cmumosi_highlevel["glove_vectors"].data.keys())[0]

In [ ]:
#The next line is not needed for standard datasets as they are all sorted based on intervals in computational sequence entries
cmumosi_highlevel.sort()

In [12]:
#Aligning to the words to get word-level alignments
cmumosi_highlevel.align('glove_vectors',collapse_functions=[myavg])
cmumosi_highlevel.impute('glove_vectors')

[2021-10-14 09:16:22.605] | Status  | Unify was called ...
[2021-10-14 09:16:22.605] | Warning | c5xsKMxpXnc entry is not shared among all sequences, removing it ...
[2021-10-14 09:16:22.606] | Success | Unify completed ...
[2021-10-14 09:16:22.606] | Status  | Pre-alignment based on <glove_vectors> computational sequence started ...
[2021-10-14 09:16:23.548] | Status  | Pre-alignment done for <FACET_4.1> ...
[2021-10-14 09:16:31.326] | Status  | Pre-alignment done for <OpenFace_2> ...
[2021-10-14 09:16:32.413] | Status  | Pre-alignment done for <b'OpenSMILE'> ...
[2021-10-14 09:16:32.566] | Status  | Pre-alignment done for <OpenSmile_emobase2010> ...
[2021-10-14 09:16:36.648] | Status  | Pre-alignment done for <COVAREP> ...
[2021-10-14 09:16:36.695] | Status  | Pre-alignment done for <Opinion Segment Labels> ...
[2021-10-14 09:16:42.568] | Status  | Pre-alignment done for <OpenFace_1> ...
[2021-10-14 09:16:43.492] | Status  | Pre-alignment done for <FACET_4.2> ...


Overall Progress:   0%|          | 0/92 [00:00<?, ? Computational Sequence Entries/s]

[2021-10-14 09:16:43.774] | Status  | Alignment starting ...


[2021-10-14 09:19:31.785] | Success | Alignment to <glove_vectors> complete.
[2021-10-14 09:19:31.785] | Status  | Replacing dataset content with aligned computational sequences
[2021-10-14 09:19:31.864] | Success | Initialized empty <OpenFace_1> computational sequence.
[2021-10-14 09:19:31.864] | Status  | Checking the format of the data in <OpenFace_1> computational sequence ...


  0%|          | 0/31923 [00:00<?, ? Computational Sequence Entries/s]

[2021-10-14 09:19:31.988] | Success | <OpenFace_1> computational sequence data in correct format.
[2021-10-14 09:19:31.989] | Status  | Checking the format of the metadata in <OpenFace_1> computational sequence ...
[2021-10-14 09:19:31.989] | Warning | <OpenFace_1> computational sequence does not have all the required metadata ... continuing 
[2021-10-14 09:19:31.989] | Success | Initialized empty <FACET_4.2> computational sequence.
[2021-10-14 09:19:31.989] | Status  | Checking the format of the data in <FACET_4.2> computational sequence ...
[2021-10-14 09:19:32.115] | Success | <FACET_4.2> computational sequence data in correct format.
[2021-10-14 09:19:32.115] | Status  | Checking the format of the metadata in <FACET_4.2> computational sequence ...
[2021-10-14 09:19:32.115] | Warning | <FACET_4.2> computational sequence does not have all the required metadata ... continuing 
[2021-10-14 09:19:32.115] | Success | Initialized empty <OpenSmile_emobase2010> computational sequence.
[2021

  0%|          | 0/49242 [00:00<?, ? Computational Sequence Entries/s]

[2021-10-14 09:19:32.205] | Success | <OpenSmile_emobase2010> computational sequence data in correct format.
[2021-10-14 09:19:32.205] | Status  | Checking the format of the metadata in <OpenSmile_emobase2010> computational sequence ...
[2021-10-14 09:19:32.205] | Warning | <OpenSmile_emobase2010> computational sequence does not have all the required metadata ... continuing 
[2021-10-14 09:19:32.205] | Success | Initialized empty <glove_vectors> computational sequence.
[2021-10-14 09:19:32.205] | Status  | Checking the format of the data in <glove_vectors> computational sequence ...
[2021-10-14 09:19:32.335] | Success | <glove_vectors> computational sequence data in correct format.
[2021-10-14 09:19:32.335] | Status  | Checking the format of the metadata in <glove_vectors> computational sequence ...
[2021-10-14 09:19:32.335] | Warning | <glove_vectors> computational sequence does not have all the required metadata ... continuing 
[2021-10-14 09:19:32.335] | Success | Initialized empty 

  0%|          | 0/49261 [00:00<?, ? Computational Sequence Entries/s]

[2021-10-14 09:19:32.471] | Success | <FACET_4.1> computational sequence data in correct format.
[2021-10-14 09:19:32.471] | Status  | Checking the format of the metadata in <FACET_4.1> computational sequence ...
[2021-10-14 09:19:32.471] | Warning | <FACET_4.1> computational sequence does not have all the required metadata ... continuing 
[2021-10-14 09:19:32.471] | Success | Initialized empty <COVAREP> computational sequence.
[2021-10-14 09:19:32.471] | Status  | Checking the format of the data in <COVAREP> computational sequence ...
[2021-10-14 09:19:32.600] | Success | <COVAREP> computational sequence data in correct format.
[2021-10-14 09:19:32.600] | Status  | Checking the format of the metadata in <COVAREP> computational sequence ...
[2021-10-14 09:19:32.600] | Warning | <COVAREP> computational sequence does not have all the required metadata ... continuing 
[2021-10-14 09:19:32.600] | Success | Initialized empty <b'OpenSMILE'> computational sequence.
[2021-10-14 09:19:32.600] |

Imputation Progress:   0%|          | 0/49261 [00:00<?, ? Reference Computational Sequence Entries/s]

[2021-10-14 09:19:32.729] | Success | <b'OpenSMILE'> computational sequence data in correct format.
[2021-10-14 09:19:32.729] | Status  | Checking the format of the metadata in <b'OpenSMILE'> computational sequence ...
[2021-10-14 09:19:32.729] | Warning | <b'OpenSMILE'> computational sequence does not have all the required metadata ... continuing 
[2021-10-14 09:19:32.729] | Success | Initialized empty <OpenFace_2> computational sequence.
[2021-10-14 09:19:32.729] | Status  | Checking the format of the data in <OpenFace_2> computational sequence ...
[2021-10-14 09:19:32.731] | Success | <OpenFace_2> computational sequence data in correct format.
[2021-10-14 09:19:32.731] | Status  | Checking the format of the metadata in <OpenFace_2> computational sequence ...
[2021-10-14 09:19:32.731] | Warning | <OpenFace_2> computational sequence does not have all the required metadata ... continuing 
[2021-10-14 09:19:32.731] | Success | Initialized empty <Opinion Segment Labels> computational seq

[2021-10-14 09:19:34.135] | Success | Imputation completed ...


In [13]:
#get the intervals and features accompanying the 100th word in the some_video
some_video_100th_word=some_video+'[100]'
for compseq_name in list(cmumosi_highlevel.computational_sequences.keys()):
	compseq=cmumosi_highlevel[compseq_name]
	print (compseq_name)
	print (numpy.array(compseq.data[some_video_100th_word]["intervals"]).shape,numpy.array(compseq.data[some_video_100th_word]["features"]).shape)
	print ("-------")

OpenFace_1
(1, 2) (1, 430)
-------
FACET_4.2
(1, 2) (1, 35)
-------
OpenSmile_emobase2010
(1, 2) (1, 1585)
-------
glove_vectors
(1, 2) (1, 300)
-------
FACET_4.1
(1, 2) (1, 47)
-------
COVAREP
(1, 2) (1, 74)
-------
b'OpenSMILE'
(1, 2) (1, 384)
-------
OpenFace_2
(1, 2) (1, 713)
-------
Opinion Segment Labels
(1, 2) (1, 1)
-------


k번째 비디오의 100번째 단어의 computational sequence를 관찰한 결과,

* Text - glove_vectors (feature: 300)

* Image - OpenFace_1 (430), FACET_4.2 (35), FACET_4.1 (47), COVAREP(74), OpenFace_2 (713)

* Audio - OpenSmile_emobase2010 (1585), b'OpenSMILE (384)

이를 통해 embedding feature 수 파악 가능.

*참고로 k번째 비디오는 glove_vector로 리스트업 하여 추출한 것.*

```python
#some random video from cmumosi_highlevel
some_video=list(cmumosi_highlevel["glove_vectors"].data.keys())[0]
```

![computational sequence](examples/compseq.png)

2번째 비디오를 관찰해보아도 마찬가지의 결과

In [15]:
#get the intervals and features accompanying the 2nd in some_video
some_video_2nd_segment=some_video+'[2]'
for compseq_name in list(cmumosi_highlevel.computational_sequences.keys()):
	compseq=cmumosi_highlevel[compseq_name]
	print (compseq_name)
	print (numpy.array(compseq.data[some_video_2nd_segment]["intervals"]).shape,numpy.array(compseq.data[some_video_2nd_segment]["features"]).shape)
	print ("-------")

OpenFace_1
(1, 2) (1, 430)
-------
FACET_4.2
(1, 2) (1, 35)
-------
OpenSmile_emobase2010
(1, 2) (1, 1585)
-------
glove_vectors
(1, 2) (1, 300)
-------
FACET_4.1
(1, 2) (1, 47)
-------
COVAREP
(1, 2) (1, 74)
-------
b'OpenSMILE'
(1, 2) (1, 384)
-------
OpenFace_2
(1, 2) (1, 713)
-------
Opinion Segment Labels
(1, 2) (1, 1)
-------


In [17]:
#Deploying the files to the disk and reading them again - Building machine learning models start right after this. No need to do alignment multiple times since aligned files can be deployed and used again.
# deploy_files={x:x for x in cmumosi_highlevel.computational_sequences.keys()}
# cmumosi_highlevel.deploy("./deployed",deploy_files)

#Reading the dumped file can be as easy as just calling the mmdataset on the deployed folder
# aligned_cmumosi_highlevel=mmdatasdk.mmdataset('./deployed')

#Now let's get the tensor ready for ML - right here we just get everything into ML ready tensors. But you should split the aligned_cmumosi_highlevel based on the standard CMU MOSI folds
#get the standard folds using mmsdk.mmdatasdk.cmu_mosi.standard_folds.standard_x_fold for x={train,test,valid}


Deployed files(파이썬 배포 csd파일) 저장 완료 상태.

In [18]:
tensors=cmumosi_highlevel.get_tensors(seq_len=25,non_sequences=["Opinion Segment Labels"],direction=False,folds=[mmdatasdk.cmu_mosi.standard_folds.standard_train_fold,mmdatasdk.cmu_mosi.standard_folds.standard_valid_fold,mmdatasdk.cmu_mosi.standard_folds.standard_test_fold])

[2021-10-14 09:49:24.785] | Status  | Hard unify was called ...
[2021-10-14 09:49:25.133] | Success | Hard unify completed ...


In [ ]:
fold_names=["train","valid","test"]

Error: Session cannot generate requests